### PR TITLE
Make Bot Cache state public

### PR DESCRIPTION
**Pull request by sbacha**
## Description
In order to properly manage internal variable state in memory, we override some of the data access in derived versions of memory scopes. Since parts of the memory management system have a strict dependency on CachedBotState, we would like to instantiate CachedBotState from our extensions.

## Specific Changes
Make the CachedBotState constructor **public**

### CODE REVIEW

**Comment by gabog** [comment]
@tomlm, @Stevenic, can you guys review this PR? I'm not familiar with the use of cachedbotstate and why we decide it to make the constructor internal.

**Comment by sgellock** [comment]
Flagging as P0 as this needs to go into 4.10.4 for PVA

**Comment by Kaiqb** [comment]
@gabog - any update on this one?

File: libraries/Microsoft.Bot.Builder/BotState.cs [hunk 1]
```diff
@@ -339,7 +339,11 @@ protected Task SetPropertyValueAsync(ITurnContext turnContext, string propertyNa
         public class CachedBotState
 #pragma warning restore CA1034 // Nested types should not be visible
         {
-            internal CachedBotState(IDictionary<string, object> state = null)
+            /// <summary>
```
Comment (gabog) [inline libraries/Microsoft.Bot.Builder/BotState.cs:342]: If we go ahead with making this class public we should make it not nested, put it in a separate file and remove the CA1034 pragma.

**Review by gabog — CHANGES_REQUESTED** [review]
Please move the class to its own file and don't have it nested if we make it public.

File: libraries/Microsoft.Bot.Builder/BotState.cs (hunk 1 above)
Comment (gabog) [reply libraries/Microsoft.Bot.Builder/BotState.cs:342]: Never mind, we can't move it outside, it is already public...

**Review by tomlm — APPROVED** [review]
:shipit:

**Review by carlosscastro — APPROVED** [review]

**Review by gabog — APPROVED** [review]
@tomlm and @carlosscastro approved so I think this good

**Pull request merged**